popsycle/synthetic.py

```diff
@@ -3512,3 +3527,390 @@ def generate_slurm_scripts(slurm_config_filename, popsycle_config_filename,
+        if type(r_max) != int:
+            raise Exception('r_max must be a float or an integer.')
+
+    if type(c) != int:
+        if type(c) != float:
+            raise Exception('c must be an integer or a float.')
+
+    if type(r_vir) != float:
+        if type(r_vir) != int:
+            raise Exception('r_vir must be a float or an integer.')
+
+    if seed is not None:
+        if type(seed) != int:
+            raise Exception('seed must be an integer.')
+
+    ##########
+    # Start of code
+    #########
+
+    # Set random seed
+    np.random.seed(seed)
+
+    t0 = time.time()
+
+    #Define parameters for NFW profile calculations
+    r_s = r_vir/c #kpc, scale radius
+    g = 4.3*(10**-3) #(pc*km^2)/(Msun*s^2)
+    h = 70 #km/(s*Mpc)
+
+    #Read in the hdf5 file that doesn't have PBHs. Product of perform_pop_syn.
+    no_pbh_hdf5_file = h5py.File(hdf5_file)
+    key_list = list(no_pbh_hdf5_file)
+    key_list = key_list[:len(key_list)-2]
+    lat_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['lat_bin_edges']))
+    long_bin = pd.DataFrame(np.array(no_pbh_hdf5_file['long_bin_edges']))
+    no_pbh_hdf5_file.close()
+
+    field_of_view_area = 2*((area/np.pi)**(1/2))
+
+    #NFW Profile calculations to determine mass of dark matter within given distance of galactic center
+    rho_crit = ((3*(h**2))/(8*np.pi*g))*(10**6) #Msun/Mpc^3
+    rho_knot = (200/3)*(rho_crit)*((c**3)/(np.log(c+1)-(c/(1+c)))) # Msun/Mpc^3
+    mass_8_kpc = (4*np.pi*rho_knot*(r_s**3)*(np.log((r_s + r_max)/r_s) - (r_max/(r_s+r_max))))*(((10**3)**3)/((10**6)**3))*fdm #Msun
+
+    #Determine the number of PBHs within that distance
+    num_pbh_8kpc = (mass_8_kpc/pbh_mass)
+    num_pbh_8kpc = round(num_pbh_8kpc, -1)
+
+    #Defining needed functions from the python package "NFWdist".
+    #Used to calculate PBH radii from galactic center assuming NFW profile.
+    #https://github.com/CullanHowlett/NFWdist
+    def pnfwunorm(q, con=5):
+        if hasattr(con, '__len__'):
+            y = np.outer(q,con)
+        else:
+            y = q*con
+        return np.log(1.0 + y)-y/(1.0 + y)
+
+    def pnfw(q, con=5, logp=False):
+        p = pnfwunorm(q, con=con)/pnfwunorm(1, con=con)
+        if hasattr(q, '__len__'):
+            p[q>1] = 1
+            p[q<=0] = 0
+        else:
+            if (q > 1):
+                p = 1
+            elif (q <= 0):
+                p = 0
+        if(logp):
+            return np.log(p)
+        else:
+            return p
+
+    def qnfw(p, con=5, logp=False):
+        if (logp):
+            p = np.exp(p)
+        if hasattr(p, '__len__'):
+            p[p>1] = 1
+            p[p<=0] = 0
+        else:
+            if (p > 1):
+                p = 1
+            elif (p <= 0):
+                p = 0
+        if hasattr(con, '__len__'):
+            p = np.outer(p,pnfwunorm(1, con=con))
+        else:
+            p *= pnfwunorm(1, con=con)
+        return (-(1.0/np.real(special.lambertw(-np.exp(-p-1))))-1)/con
```

Comment:
Please label these with more comments. I do not understand what these functions are doing or how they work.